Tests/SentryTests/Profiling/SentryProfilerSwiftTests.swift

```diff
@@ -0,0 +1,373 @@
+        let expectedEnvironment = "test-environment"
+        fixture.options.environment = expectedEnvironment
+        performTest(transactionEnvironment: expectedEnvironment)
+    }
+
+    func testProfilingDataContainsEnvironmentSetFromConfigureScope() {
+        fixture.options.profilesSampleRate = 1.0
+        fixture.options.tracesSampleRate = 1.0
+        let expectedEnvironment = "test-environment"
+        fixture.hub.configureScope { scope in
+            scope.setEnvironment(expectedEnvironment)
+        }
+        performTest(transactionEnvironment: expectedEnvironment)
+    }
+
+    func testStartTransaction_NotSamplingProfileUsingEnableProfiling() {
+        assertProfilesSampler(expectedDecision: .no) { options in
+            options.enableProfiling_DEPRECATED_TEST_ONLY = false
+        }
+    }
+
+    func testStartTransaction_SamplingProfileUsingEnableProfiling() {
+        assertProfilesSampler(expectedDecision: .yes) { options in
+            options.enableProfiling_DEPRECATED_TEST_ONLY = true
+        }
+    }
+
+    func testStartTransaction_NotSamplingProfileUsingSampleRate() {
+        assertProfilesSampler(expectedDecision: .no) { options in
+            options.profilesSampleRate = 0.49
+        }
+    }
+
+    func testStartTransaction_SamplingProfileUsingSampleRate() {
+        assertProfilesSampler(expectedDecision: .yes) { options in
+            options.profilesSampleRate = 0.5
+        }
+    }
+
+    func testStartTransaction_SamplingProfileUsingProfilesSampler() {
+        assertProfilesSampler(expectedDecision: .yes) { options in
+            options.profilesSampler = { _ in return 0.51 }
+        }
+    }
+
+    func testStartTransaction_WhenProfilesSampleRateAndProfilesSamplerNil() {
+        assertProfilesSampler(expectedDecision: .no) { options in
+            options.profilesSampleRate = nil
+            options.profilesSampler = { _ in return nil }
+        }
+    }
+
+    func testStartTransaction_WhenProfilesSamplerOutOfRange_TooBig() {
+        assertProfilesSampler(expectedDecision: .no) { options in
+            options.profilesSampler = { _ in return 1.01 }
+        }
+    }
+
+    func testStartTransaction_WhenProfilesSamplersOutOfRange_TooSmall() {
+        assertProfilesSampler(expectedDecision: .no) { options in
+            options.profilesSampler = { _ in return -0.01 }
+        }
+    }
+}
+
+private extension SentryProfilerSwiftTests {
+    func performTest(transactionEnvironment: String = kSentryDefaultEnvironment, duration: TimeInterval = 2.0, numberOfTransactions: Int = 1, customAssertions: (([String: Any]) -> Void)? = nil) {
+        let profileExpectation = expectation(description: "collects profiling data")
+        let span = fixture.hub.startTransaction(name: fixture.transactionName, operation: fixture.transactionOperation)
+        // Give it time to collect a profile, otherwise there will be no samples.
+        DispatchQueue.global().asyncAfter(deadline: .now() + duration) {
```

Comment:
`m`: Those delays are often the cause of flaky tests. I think we should put some effort into adding some abstractions to the code under tests so we don't need those delays. If you don't understand what I'm referring to, I can elaborate on it, but that is some effort, so I ask before I do that.

Reply:
Are you talking about abstractions besides `TestCurrentDateProvider`, like mocking out parts of `SentryProfiler`/`SentryTracer`? Currently these tests look at actual frame info that is gathered from the profiler in `assertValidProfileData`. That is why this code is here: https://github.com/getsentry/sentry-cocoa/pull/2105/files#diff-b341a165bb5e23dbf87f2cbbcd39bdf0dbef45c3e57a46e8cda4d49bd4019a0eR248-R253. We could instead mock something to provide that data, but then we would no longer be actually testing this part of the profiler.

Reply:
In any case, these tests have worked like this for a while, so let's leave this for a separate PR.